pallets/bridge/src/lib.rs

```diff
@@ -0,0 +1,621 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod hashing;
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::{Decode, Encode, EncodeLike};
+	pub use frame_support::{
+		pallet_prelude::*, traits::StorageVersion, weights::GetDispatchInfo, PalletId, Parameter,
+	};
+	use frame_system::{
+		pallet_prelude::*,
+		{self as system},
+	};
+	use scale_info::TypeInfo;
+	pub use sp_core::U256;
+	use sp_runtime::{
+		traits::{AccountIdConversion, Dispatchable},
+		RuntimeDebug,
+	};
+	use sp_std::prelude::*;
+
+	const DEFAULT_RELAYER_THRESHOLD: u32 = 1;
+	const MODULE_ID: PalletId = PalletId(*b"litry/bg");
+
+	pub type BridgeChainId = u8;
+	pub type DepositNonce = u64;
+	pub type ResourceId = [u8; 32];
+
+	/// Helper function to concatenate a chain ID and some bytes to produce a resource ID.
+	/// The common format is (31 bytes unique ID + 1 byte chain ID).
+	pub fn derive_resource_id(chain: u8, id: &[u8]) -> ResourceId {
+		let mut r_id: ResourceId = [0; 32];
+		r_id[31] = chain; // last byte is chain id
+		let range = if id.len() > 31 { 31 } else { id.len() }; // Use at most 31 bytes
+		for i in 0..range {
+			r_id[30 - i] = id[range - 1 - i]; // Ensure left padding for eth compatibility
+		}
+		r_id
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum ProposalStatus {
+		Initiated,
+		Approved,
+		Rejected,
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub struct ProposalVotes<AccountId, BlockNumber> {
+		pub votes_for: Vec<AccountId>,
+		pub votes_against: Vec<AccountId>,
+		pub status: ProposalStatus,
+		pub expiry: BlockNumber,
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum BridgeEvent {
+		FungibleTransfer(BridgeChainId, DepositNonce, ResourceId, U256, Vec<u8>),
+		NonFungibleTransfer(BridgeChainId, DepositNonce, ResourceId, Vec<u8>, Vec<u8>, Vec<u8>),
+		GenericTransfer(BridgeChainId, DepositNonce, ResourceId, Vec<u8>),
+	}
+
+	impl<A: PartialEq, B: PartialOrd + Default> ProposalVotes<A, B> {
+		/// Attempts to mark the proposal as approve or rejected.
+		/// Returns true if the status changes from active.
+		pub fn try_to_complete(&mut self, threshold: u32, total: u32) -> ProposalStatus {
```

Comment:
TBH both parameters seem strange to me:
- `threshold` appears to act as both the passingThreshold and totalThreshold
- `total` shouldn't it be calculated from `votes_for.len() + votes_against.len()`? If it's meant to be "maximum possible total votes", then some sanity checks should be performed, e.g. `votes_for.len() + votes_against.len() <= total`

Reply:
- threshold is dynamic, maybe updated at runtime
- some node may miss the vote, total != votes_for.len() + votes_against.len(), and vote could be finalized before all nodes vote
- votes_for.len() + votes_against.len() <= total not needed.

Reply:
I'm only checking this fn alone without examining which values the parameters can be. I believe the code will work as expected in the majority of cases though